### PR TITLE
Nullable annotations for metadata 2

### PR DESCRIPTION
**Pull request by roji**
Am just going to drop this here, sorry @AndriySvyryd... This annotates Model, Property, Navigation, Key, Index, and all their extensions.

* ISkipNavigation.ForeignKey and Inverse are non-nullable (ModelValidator ensures this), but these are nullable on IMutableSkipNavigation and IConventionSkipNavigation. Seems like a pretty satisfactory solution, no?
* ~IPropertyBase.ClrType is nullable (because of skip navigations), but IProperty overrides it to be non-nullable.~
* Changes a bit of query too (/cc @smitpatel)

Part of #19007

### CODE REVIEW

File: src/EFCore/Metadata/Internal/ForeignKey.cs [hunk 1]
```diff
@@ -77,6 +79,11 @@ public class ForeignKey : ConventionAnnotatable, IMutableForeignKey, IConvention
             }
 
             Builder = new InternalForeignKeyBuilder(this, dependentEntityType.Model.Builder);
+
+            // TODO-NULLABLE: the following seem to be externally-set - we can consider making them nullable, or just leave like this
```
Comment (roji) [inline src/EFCore/Metadata/Internal/ForeignKey.cs:83]: /cc @AndriySvyryd

File: src/EFCore/Metadata/Internal/Navigation.cs [hunk 2]
```diff
@@ -181,8 +183,9 @@ public virtual void SetIsEagerLoaded(bool? eagerLoaded, ConfigurationSource conf
                 return false;
             }
 
+            // TODO-NULLABLE: We check for targetType.ClrType being null above, but not for sourceType
```
Comment (roji) [inline src/EFCore/Metadata/Internal/Navigation.cs:186]: /cc @AndriySvyryd

File: src/EFCore/Extensions/PropertyExtensions.cs [hunk 3]
```diff
@@ -357,7 +359,8 @@ public static ValueComparer GetStructuralValueComparer([NotNull] this IProperty
         /// <returns> A new equality comparer. </returns>
         public static IEqualityComparer<TProperty> CreateKeyEqualityComparer<TProperty>([NotNull] this IProperty property)
         {
-            var comparer = property.GetKeyValueComparer();
+            // TODO-NULLABLE: GetKeyValueComparer can return null, bug?
```
Comment (roji) [inline src/EFCore/Extensions/PropertyExtensions.cs:362]: /cc @AndriySvyryd

File: src/EFCore/Metadata/Internal/ForeignKey.cs [hunk 4]
```diff
@@ -526,21 +533,23 @@ public virtual bool IsUnique
                     CoreStrings.NonUniqueRequiredDependentForeignKey(Properties.Format(), DeclaringEntityType.DisplayName()));
             }
 
+            // TODO-NULLABLE: Should we check DeclaringEntityType.ClrType for null as well?
```
Comment (roji) [inline src/EFCore/Metadata/Internal/ForeignKey.cs:536]: /cc @AndriySvyryd

File: src/EFCore/Metadata/Internal/ForeignKey.cs [hunk 5]
```diff
@@ -526,21 +533,23 @@ public virtual bool IsUnique
...
             {
                 if (!Internal.Navigation.IsCompatible(
                     PrincipalToDependent.GetIdentifyingMemberInfo(),
                     PrincipalEntityType.ClrType,
-                    DeclaringEntityType.ClrType,
+                    DeclaringEntityType.ClrType!,
                     !unique,
                     shouldThrow: false))
                 {
+                    // TODO-NULLABLE: Bug if PropertyInfo is null (FieldInfo?)
```
Comment (roji) [inline src/EFCore/Metadata/Internal/ForeignKey.cs:548]: /cc @AndriySvyryd

File: src/EFCore/Metadata/Internal/ForeignKey.cs [hunk 6]
```diff
@@ -1280,11 +1289,12 @@ public override string ToString()
                 return false;
             }
 
+            // TODO-NULLABLE: add entry check above against shadow entities?
```
Comment (roji) [inline src/EFCore/Metadata/Internal/ForeignKey.cs:1292]: /cc @AndriySvyryd

File: src/EFCore/Metadata/Internal/Property.cs [hunk 7]
```diff
@@ -485,9 +489,10 @@ public virtual CoreTypeMapping TypeMapping
         ///     any release. You should only use it directly in your code with extreme caution and knowing that
         ///     doing so can result in application failures when updating to a new Entity Framework Core release.
         /// </summary>
-        public virtual CoreTypeMapping SetTypeMapping([NotNull] CoreTypeMapping typeMapping, ConfigurationSource configurationSource)
+        public virtual CoreTypeMapping? SetTypeMapping([NotNull] CoreTypeMapping typeMapping, ConfigurationSource configurationSource)
         {
             _typeMapping = typeMapping;
+            // TODO-NULLABLE: null check conflicts with [NotNull] annotation above, which seems like it may be incorrect.
```
Comment (roji) [inline src/EFCore/Metadata/Internal/Property.cs:495]: /cc @AndriySvyryd

File: src/EFCore/Metadata/Internal/ServiceProperty.cs [hunk 8]
```diff
@@ -31,16 +33,18 @@ public class ServiceProperty : PropertyBase, IMutableServiceProperty, IConventio
...
+            [CanBeNull] FieldInfo? fieldInfo,
             [NotNull] EntityType declaringEntityType,
             ConfigurationSource configurationSource)
             : base(name, propertyInfo, fieldInfo, configurationSource)
         {
             Check.NotNull(declaringEntityType, nameof(declaringEntityType));
 
             DeclaringEntityType = declaringEntityType;
-            ClrType = propertyInfo?.PropertyType ?? fieldInfo?.FieldType;
+            // TODO-NULLABLE: are we missing a check that either propertyInfo or fieldInfo are given? Or are shadow service properties
```
Comment (roji) [inline src/EFCore/Metadata/Internal/ServiceProperty.cs:45]: /cc @AndriySvyryd

File: src/EFCore/Metadata/Internal/SkipNavigation.cs [hunk 9]
```diff
@@ -159,23 +161,21 @@ public virtual EntityType JoinEntityType
...
+        public virtual ForeignKey? SetForeignKey([CanBeNull] ForeignKey? foreignKey, ConfigurationSource configurationSource)
         {
             var oldForeignKey = ForeignKey;
             var isChanging = foreignKey != ForeignKey;
 
-            if (oldForeignKey != null)
-            {
-                oldForeignKey.ReferencingSkipNavigations.Remove(this);
-            }
+            // TODO-NULLABLE: Need to check for null ReferencingSkipNavigations?
```
Comment (roji) [inline src/EFCore/Metadata/Internal/SkipNavigation.cs:169]: /cc @AndriySvyryd

File: src/EFCore/Metadata/Internal/SkipNavigation.cs [hunk 10]
```diff
@@ -197,10 +197,11 @@ public virtual ForeignKey SetForeignKey([CanBeNull] ForeignKey foreignKey, Confi
             if (Inverse?.JoinEntityType != null
                 && Inverse.JoinEntityType != JoinEntityType)
             {
+                // TODO-NULLABLE: JoinEntityType can be null?
```
Comment (roji) [inline src/EFCore/Metadata/Internal/SkipNavigation.cs:200]: /cc @AndriySvyryd

File: src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.Expressions.cs [hunk 11]
```diff
@@ -433,7 +433,7 @@ protected override Expression VisitChildren(ExpressionVisitor visitor)
             public INavigation Navigation { get; }
             public EntityReference EntityReference { get; }
 
-            public override Type Type
+            public override Type? Type
```
Comment (smitpatel) [inline src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.Expressions.cs:436]: this is non-nullable. Navigation.ClrType won't be null at runtime.

File: src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs [hunk 12]
```diff
@@ -1645,7 +1645,7 @@ private static IEnumerable<INavigationBase> FindNavigations(IEntityType entityTy
             }
             else
             {
-                foreach (var derivedNavigation in entityType.GetDerivedTypes()
+                foreach (var derivedNavigation in (IEnumerable<INavigation>)entityType.GetDerivedTypes()
```
Comment (smitpatel) [inline src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs:1648]: Just get null check out like skip part.

File: src/EFCore/Query/MaterializeCollectionNavigationExpression.cs [hunk 13]
```diff
@@ -51,7 +51,7 @@ public sealed override ExpressionType NodeType
             => ExpressionType.Extension;
 
         /// <inheritdoc />
-        public override Type Type
+        public override Type? Type
```
Comment (smitpatel) [inline src/EFCore/Query/MaterializeCollectionNavigationExpression.cs:54]: This is non-nullable as mentioned above.

File: src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs (hunk 12 above)
Comment (roji) [reply src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs:1648]: OK. FWIW filtering out nulls this is a specific case they're looking into for C# 10 (https://github.com/dotnet/csharplang/issues/3951)

File: src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.Expressions.cs (hunk 11 above)
Comment (roji) [reply src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.Expressions.cs:436]: Fixing on IPropertyBase side as discussed with @AndriySvyryd

**Comment by roji** [comment]
@smitpatel addressed your comments

File: src/EFCore/Extensions/MutableEntityTypeExtensions.cs [hunk 14]
```diff
@@ -179,7 +181,7 @@ public static IMutableKey FindKey([NotNull] this IMutableEntityType entityType,
         /// <param name="entityType"> The entity type. </param>
         /// <param name="property"> The property to use as an alternate key. </param>
         /// <returns> The newly created key. </returns>
-        public static IMutableKey AddKey(
+        public static IMutableKey? AddKey(
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/MutableEntityTypeExtensions.cs:184]: Non-nullable

File: src/EFCore/Extensions/MutableEntityTypeExtensions.cs [hunk 15]
```diff
@@ -314,7 +316,7 @@ public static IEnumerable<IMutableForeignKey> GetDeclaredReferencingForeignKeys(
         ///     base type of the hierarchy).
         /// </param>
         /// <returns> The newly created foreign key. </returns>
-        public static IMutableForeignKey AddForeignKey(
+        public static IMutableForeignKey? AddForeignKey(
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/MutableEntityTypeExtensions.cs:319]: Non-nullable

File: src/EFCore/Extensions/MutableEntityTypeExtensions.cs [hunk 16]
```diff
@@ -451,7 +453,7 @@ public static IMutableProperty FindDeclaredProperty([NotNull] this IMutableEntit
         /// <param name="entityType"> The entity type. </param>
         /// <param name="memberInfo"> The corresponding member on the entity class. </param>
         /// <returns> The newly created property. </returns>
-        public static IMutableProperty AddProperty(
+        public static IMutableProperty? AddProperty(
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/MutableEntityTypeExtensions.cs:456]: Non-nullable here and for the resto of the Add* methods in this file

File: src/EFCore/Extensions/MutableModelExtensions.cs [hunk 17]
```diff
@@ -204,7 +206,7 @@ public static bool IsIgnored([NotNull] this IMutableModel model, [NotNull] Type
         /// <param name="model"> The model to add the ignored type to. </param>
         /// <param name="type"> The entity type to be ignored. </param>
         /// <returns> The name of the ignored type. </returns>
-        public static string AddIgnored([NotNull] this IMutableModel model, [NotNull] Type type)
+        public static string? AddIgnored([NotNull] this IMutableModel model, [NotNull] Type type)
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/MutableModelExtensions.cs:209]: Non-nullable

File: src/EFCore/Extensions/MutableModelExtensions.cs [hunk 18]
```diff
@@ -260,7 +262,7 @@ public static void AddShared([NotNull] this IMutableModel model, [NotNull] Type
         /// </summary>
         /// <param name="model"> The model to finalize. </param>
         /// <returns> The finalized <see cref="IModel" />. </returns>
-        public static IModel FinalizeModel([NotNull] this IMutableModel model)
+        public static IModel? FinalizeModel([NotNull] this IMutableModel model)
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/MutableModelExtensions.cs:265]: Non-nullable

File: src/EFCore/Extensions/MutablePropertyExtensions.cs [hunk 19]
```diff
@@ -189,23 +191,23 @@ public static void SetAfterSaveBehavior([NotNull] this IMutableProperty property
...
+        public static void SetProviderClrType([NotNull] this IMutableProperty property, [CanBeNull] Type? providerClrType)
             => property.AsProperty().SetProviderClrType(providerClrType, ConfigurationSource.Explicit);
 
         /// <summary>
         ///     Sets the <see cref="CoreTypeMapping" /> for the given property
         /// </summary>
         /// <param name="property"> The property. </param>
         /// <param name="typeMapping"> The <see cref="CoreTypeMapping" /> for this property. </param>
-        public static CoreTypeMapping SetTypeMapping(
+        public static CoreTypeMapping? SetTypeMapping(
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/MutablePropertyExtensions.cs:210]: Non-nullable

File: src/EFCore/Extensions/PropertyExtensions.cs (hunk 3 above)
Comment (AndriySvyryd) [reply src/EFCore/Extensions/PropertyExtensions.cs:362]: This **is** about model-building vs runtime. I'll fix this and other methods that depend on type mapping in https://github.com/dotnet/efcore/issues/22031

File: src/EFCore/Extensions/MutableEntityTypeExtensions.cs (hunk 14 above)
Comment (roji) [reply src/EFCore/Extensions/MutableEntityTypeExtensions.cs:184]: Left those out, sorry - will correct.

File: src/EFCore/Extensions/MutableModelExtensions.cs (hunk 18 above)
Comment (roji) [reply src/EFCore/Extensions/MutableModelExtensions.cs:265]: Have done this like the Add models (by leaving FinalizeModel nullable on Model and putting a bang in the extension over IMutableModel)

Comment (AndriySvyryd) [reply src/EFCore/Extensions/MutableModelExtensions.cs]: It might be out-of-scope for this PR, but `Model.FinalizeModel` should return `null`, that's just silly

Comment (roji) [reply src/EFCore/Extensions/MutableModelExtensions.cs]: If that's so, I can just make that change. I was assuming that anything that comes out of ConventionDispatcher may be nullable, but this indeed doesn't make much sense here.

File: src/EFCore/ChangeTracking/Internal/KeyValueFactoryFactory.cs [hunk 20]
```diff
@@ -28,7 +28,6 @@ public virtual IPrincipalKeyValueFactory<TKey> Create<TKey>([NotNull] IKey key)
                 ? CreateSimpleFactory<TKey>(key)
                 : (IPrincipalKeyValueFactory<TKey>)CreateCompositeFactory(key);
 
-        [UsedImplicitly]
```
Comment (smitpatel) [inline src/EFCore/ChangeTracking/Internal/KeyValueFactoryFactory.cs:31]: Is this related? needed? not needed?
I thought they are used by R# to specify that this method is called via reflection and not dead.

File: src/EFCore/Extensions/MutableModelExtensions.cs (hunk 18 above)
Comment (roji) [reply src/EFCore/Extensions/MutableModelExtensions.cs]: It does seem like OnModelFinalizing can return null (we check), so although I can put a bang before returning I'm not sure what would happen in that case.

File: src/EFCore/ChangeTracking/Internal/KeyValueFactoryFactory.cs (hunk 20 above)
Comment (roji) [reply src/EFCore/ChangeTracking/Internal/KeyValueFactoryFactory.cs:31]: Just removed it because it is called (very explicitly) three lines above :) Otherwise yeah, these help R# not identify things as unused.

File: src/EFCore/Metadata/Internal/EntityTypeExtensions.cs [hunk 21]
```diff
@@ -152,7 +154,7 @@ public static IEntityType FindInOwnershipPath([NotNull] this IEntityType entityT
                 }
 
                 owner = ownership.PrincipalEntityType;
-                if (owner.ClrType.IsAssignableFrom(targetType))
+                if (owner.ClrType!.IsAssignableFrom(targetType))
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/EntityTypeExtensions.cs:157]: Add a null check until we remove shadow entity types

File: src/EFCore/Extensions/MutableModelExtensions.cs (hunk 18 above)
Comment (AndriySvyryd) [reply src/EFCore/Extensions/MutableModelExtensions.cs]: It can be `null` and we have a test where it's `null`, but in practice there's no reasonable scenario where it should be `null`

File: src/EFCore/Metadata/Internal/ForeignKey.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/ForeignKey.cs:83]: These should be nullable

File: src/EFCore/Metadata/Internal/ForeignKey.cs (hunk 4 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/ForeignKey.cs:536]: This won't be hit in practice, but we can add the check until we remove shadow entity types to make the compiler happy

File: src/EFCore/Metadata/Internal/ForeignKey.cs (hunk 5 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/ForeignKey.cs:548]: Should use `GetIdentifyingMemberInfo()` instead

Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/ForeignKey.cs:548]: Or just `Name`

File: src/EFCore/Metadata/Internal/ForeignKey.cs (hunk 6 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/ForeignKey.cs:1292]: Yes, add a check. I remember that there once was one, I guess it was refactored out.

File: src/EFCore/Metadata/Internal/ForeignKey.cs (hunk 1 above)
Comment (roji) [reply src/EFCore/Metadata/Internal/ForeignKey.cs:83]: OK, but ForeignKeyExtensions.CreateDependentsMapFactory assumes DependentsMapFactory is never null. Bang it or check and throw?

File: src/EFCore/Extensions/PropertyExtensions.cs (hunk 3 above)
Comment (roji) [reply src/EFCore/Extensions/PropertyExtensions.cs:362]: OK. Should I leave a TODO-NULLABLE with a reference to #22031, or do you prefer just leaving bangs with no comments?

File: src/EFCore/Extensions/MutableModelExtensions.cs (hunk 18 above)
Comment (roji) [reply src/EFCore/Extensions/MutableModelExtensions.cs]: OK, will bang

File: src/EFCore/Metadata/Internal/Navigation.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/Navigation.cs:186]: We check it at the call sites, but we can move it here for clarity

File: src/EFCore/Extensions/PropertyExtensions.cs (hunk 3 above)
Comment (AndriySvyryd) [reply src/EFCore/Extensions/PropertyExtensions.cs:362]: either way is fine

File: src/EFCore/Extensions/ConventionPropertyExtensions.cs [hunk 22]
```diff
@@ -88,7 +90,7 @@ public static IEnumerable<IConventionKey> GetContainingKeys([NotNull] this IConv
         /// <param name="property"> The property. </param>
         /// <param name="typeMapping"> The <see cref="CoreTypeMapping" /> for this property. </param>
         /// <param name="fromDataAnnotation"> Indicates whether the configuration was specified using a data annotation. </param>
-        public static CoreTypeMapping SetTypeMapping(
```
Comment (smitpatel) [inline src/EFCore/Extensions/ConventionPropertyExtensions.cs:91]: Is this method missing <returns> in XML docs?

File: src/EFCore/Metadata/Internal/ForeignKey.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/ForeignKey.cs:83]: Bang and reference #22031

File: src/EFCore/Extensions/MutableEntityTypeExtensions.cs [hunk 23]
```diff
@@ -156,7 +158,7 @@ public static IEnumerable<IMutableIndex> GetDeclaredIndexes([NotNull] this IMuta
         /// <param name="entityType"> The entity type. </param>
         /// <param name="name"> The name of the property to remove. </param>
         /// <returns> The property that was removed. </returns>
-        public static IMutableProperty RemoveProperty([NotNull] this IMutableEntityType entityType, [NotNull] string name)
+        public static IMutableProperty? RemoveProperty([NotNull] this IMutableEntityType entityType, [NotNull] string name)
```
Comment (smitpatel) [inline src/EFCore/Extensions/MutableEntityTypeExtensions.cs:161]: Would this be non-null? The returns does not say that null would be returned.

Comment (smitpatel) [reply src/EFCore/Extensions/MutableEntityTypeExtensions.cs:161]: Same for other Remove* methods in this class.

File: src/EFCore/Metadata/Internal/Property.cs (hunk 7 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/Property.cs:495]: The annotation is correct, the check should be removed

Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/Property.cs:495]: `HasTypeMapping` should also have `[NotNull]` and callers should add a check.

File: src/EFCore/Metadata/Internal/PropertyExtensions.cs [hunk 24]
```diff
@@ -59,7 +61,7 @@ public static IEnumerable<IForeignKey> GetReferencingForeignKeys([NotNull] this
         ///     any release. You should only use it directly in your code with extreme caution and knowing that
         ///     doing so can result in application failures when updating to a new Entity Framework Core release.
         /// </summary>
-        public static IProperty GetGenerationProperty([NotNull] this IProperty property)
+        public static IProperty? GetGenerationProperty([NotNull] this IProperty property)
```
Comment (smitpatel) [inline src/EFCore/Metadata/Internal/PropertyExtensions.cs:64]: I feel odd that Get method is returning null.

File: src/EFCore/Extensions/MutableEntityTypeExtensions.cs (hunk 23 above)
Comment (AndriySvyryd) [reply src/EFCore/Extensions/MutableEntityTypeExtensions.cs:161]: Will return null if not found. We can fix the docs

File: src/EFCore/Metadata/Internal/PropertyExtensions.cs (hunk 24 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/PropertyExtensions.cs:64]: This is Internal, so we can rename to Find

File: src/EFCore/Metadata/Internal/PropertyBase.cs [hunk 25]
```diff
@@ -24,17 +27,17 @@ namespace Microsoft.EntityFrameworkCore.Metadata.Internal
...
-        private IClrPropertySetter _setter;
-        private IClrPropertySetter _materializationSetter;
-        private PropertyAccessors _accessors;
-        private PropertyIndexes _indexes;
+        private IClrPropertyGetter? _getter;
+        private IClrPropertySetter? _setter;
+        private IClrPropertySetter? _materializationSetter;
+        private PropertyAccessors? _accessors;
+        private PropertyIndexes? _indexes;
         private ConfigurationSource _configurationSource;
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/PropertyBase.cs:39]: Move this up to `_fieldInfoConfigurationSource`

File: src/EFCore/Metadata/Internal/ServiceProperty.cs (hunk 8 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/ServiceProperty.cs:45]: We can either add a check or change to a single `MemberInfo` parameter.

File: src/EFCore/Metadata/Internal/SkipNavigation.cs (hunk 9 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/SkipNavigation.cs:169]: No, ReferencingSkipNavigations will never be `null` here.

File: src/EFCore/Metadata/Internal/SkipNavigation.cs (hunk 10 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/SkipNavigation.cs:200]: No, `ProcessForeignKey(foreignKey);` sets it.

**Review by AndriySvyryd — APPROVED** [review]

File: src/EFCore/Extensions/MutableEntityTypeExtensions.cs (hunk 23 above)
Comment (roji) [reply src/EFCore/Extensions/MutableEntityTypeExtensions.cs:161]: Fixed up the docs.

File: src/EFCore/Metadata/Internal/EntityTypeExtensions.cs (hunk 21 above)
Comment (roji) [reply src/EFCore/Metadata/Internal/EntityTypeExtensions.cs:157]: Changed bang to runtime check

File: src/EFCore/Metadata/Internal/SkipNavigation.cs (hunk 10 above)
Comment (roji) [reply src/EFCore/Metadata/Internal/SkipNavigation.cs:200]: Removed TODO-NULLABLE

File: src/EFCore/Metadata/Internal/SkipNavigation.cs (hunk 9 above)
Comment (roji) [reply src/EFCore/Metadata/Internal/SkipNavigation.cs:169]: Removed TODO-NULLABLE

File: src/EFCore/Metadata/Internal/ServiceProperty.cs (hunk 8 above)
Comment (roji) [reply src/EFCore/Metadata/Internal/ServiceProperty.cs:45]: Changing to a single MemberInfo sounds good, but it's a larger change... Let's leave for later (if we care enough). Removing the TODO-NULLABLE.

File: src/EFCore/Metadata/Internal/Property.cs (hunk 7 above)
Comment (roji) [reply src/EFCore/Metadata/Internal/Property.cs:495]: So no way to unset a type mapping, right? Making sure because the docs on HasTypeMapping explicitly document accepting null to remove type mapping.

The only usage in our codebase seems to be in TypeMappingConvention.ProcessModelFinalizing, which we'll annotate a bit later - should that throw if FindMapping returns null?

In the meantime, changed HasTypeMapping (and CanHasTypeMapping) to have [NotNull].

File: src/EFCore/Metadata/Internal/Navigation.cs (hunk 2 above)
Comment (roji) [reply src/EFCore/Metadata/Internal/Navigation.cs:186]: Added a check.

File: src/EFCore/Extensions/PropertyExtensions.cs (hunk 3 above)
Comment (roji) [reply src/EFCore/Extensions/PropertyExtensions.cs:362]: Left TODO-NULLABLE

File: src/EFCore/Extensions/ConventionPropertyExtensions.cs (hunk 22 above)
Comment (roji) [reply src/EFCore/Extensions/ConventionPropertyExtensions.cs:91]: You mean in the published docs? I think this is a new method in 5.0 so not yet published, no?

File: src/EFCore/Metadata/Internal/Property.cs (hunk 7 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/Property.cs]: I will likely obsolete these methods when #22031 is done

File: src/EFCore/Extensions/ConventionPropertyExtensions.cs (hunk 22 above)
Comment (smitpatel) [reply src/EFCore/Extensions/ConventionPropertyExtensions.cs]: In XML docs in source code. The method is not void but still does not specify <returns> tag.

Comment (roji) [reply src/EFCore/Extensions/ConventionPropertyExtensions.cs]: Right, will fix this in the next nullability PR, thanks.

File: src/EFCore/Metadata/Internal/Property.cs (hunk 7 above)
Comment (roji) [reply src/EFCore/Metadata/Internal/Property.cs]: @AndriySvyryd as pointed by @smitpatel in https://github.com/dotnet/efcore/pull/23147#discussion_r515397983, SetTypeMapping returns CoreTypeMapping but that isn't documented. Since the parameter is now [NotNull], it seems they only ever return the supplied parameter. So these can probably be changed to return void, but if you're planning to obsolete them anyway for 6.0...

File: src/EFCore/Extensions/ConventionPropertyExtensions.cs (hunk 22 above)
Comment (roji) [reply src/EFCore/Extensions/ConventionPropertyExtensions.cs]: Actually, this now always returns the provided parameter (https://github.com/dotnet/efcore/pull/23147#discussion_r515586436), I'm embarrassed to document the return type as "always identical to the parameter" :)

File: src/EFCore/Metadata/Internal/Property.cs (hunk 7 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/Property.cs]: We can say that it returns the type mapping that was set

File: src/EFCore/Extensions/ConventionPropertyExtensions.cs (hunk 22 above)
Comment (AndriySvyryd) [reply src/EFCore/Extensions/ConventionPropertyExtensions.cs]: Here too we can say that it returns the type mapping that was set
